File: apps/enterprise/src/services/sub-accounts.ts
/**
 * Sub-Account Service
 * 
 * Agency and reseller sub-account management
 */

import { Pool } from 'pg';
import Redis from 'ioredis';
import { v4 as uuidv4 } from 'uuid';
import { config, EnterprisePlan } from '../config.js';

// Result type for error handling
type Result<T, E = Error> = { ok: true; value: T } | { ok: false; error: E };

export enum SubAccountStatus {
  ACTIVE = 'active',
  SUSPENDED = 'suspended',
  PENDING = 'pending',
  DELETED = 'deleted',
}

export interface SubAccount {
  id: string;
  parentAccountId: string;
  name: string;
  slug: string;
  email: string;
  status: SubAccountStatus;
  volumeLimit: number;
  volumeUsed: number;
  settings: SubAccountSettings;
  createdAt: Date;
  updatedAt: Date;
}

export interface SubAccountSettings {
  inheritParentDomains: boolean;
  inheritParentTemplates: boolean;
  inheritParentSuppressions: boolean;
  customDomains: string[];
  apiKeyPrefix?: string;
  webhookUrl?: string;
  notificationEmails: string[];
}

export interface SubAccountStats {
  id: string;
  name: string;
  emailsSent: number;
  emailsDelivered: number;
  emailsBounced: number;
  emailsOpened: number;
  emailsClicked: number;
  volumeUsed: number;
  volumeLimit: number;
  volumePercentage: number;
}

export interface VolumeAllocation {
  subAccountId: string;
  allocatedVolume: number;
  usedVolume: number;
  remainingVolume: number;
  periodStart: Date;
  periodEnd: Date;
}

export interface SubAccountAPIKey {
  id: string;
  subAccountId: string;
  name: string;
  keyPrefix: string;
  keyHash: string;
  scopes: string[];
  expiresAt?: Date;
  createdAt: Date;
  lastUsedAt?: Date;
}

/**
 * Sub-Account Service for agency/reseller management
 */
export class SubAccountService {
  private pool: Pool;
  private redis: Redis;

  constructor(pool: Pool, redis: Redis) {
    this.pool = pool;
    this.redis = redis;
  }

  /**
   * Create a new sub-account
   */
  async createSubAccount(
    parentAccountId: string,
    data: {
      name: string;
      slug: string;
      email: string;
      volumeLimit?: number;
      settings?: Partial<SubAccountSettings>;
    }
  ): Promise<Result<SubAccount>> {
    try {
      // Check sub-account limit
      const countResult = await this.pool.query(`
        SELECT COUNT(*) as count FROM ent_sub_accounts
        WHERE parent_account_id = $1 AND status != 'deleted'
      `, [parentAccountId]);

      if (parseInt(countResult.rows[0].count, 10) >= config.subAccounts.maxSubAccounts) {
        return {
          ok: false,
          error: new Error(`Maximum sub-account limit (${config.subAccounts.maxSubAccounts}) reached`),
        };
      }

      // Check slug uniqueness
      const slugCheck = await this.pool.query(`
        SELECT id FROM ent_sub_accounts
        WHERE parent_account_id = $1 AND slug = $2 AND status != 'deleted'
      `, [parentAccountId, data.slug]);

      if (slugCheck.rows.length > 0) {
        return { ok: false, error: new Error('Sub-account slug already exists') };
      }

      const id = uuidv4();
      const settings: SubAccountSettings = {
        inheritParentDomains: config.subAccounts.inheritParentSettings,
        inheritParentTemplates: config.subAccounts.inheritParentSettings,
        inheritParentSuppressions: config.subAccounts.inheritParentSettings,
        customDomains: [],
        notificationEmails: [data.email],
        ...data.settings,
      };

      const result = await this.pool.query(`
        INSERT INTO ent_sub_accounts (
          id, parent_account_id, name, slug, email, status,
          volume_limit, volume_used, settings, created_at, updated_at
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, 0, $8, NOW(), NOW())
        RETURNING *
      `, [
        id,
        parentAccountId,
        data.name,
        data.slug,
        data.email,
        SubAccountStatus.ACTIVE,
        data.volumeLimit || 0,
        JSON.stringify(settings),
      ]);

      const row = result.rows[0];
      return {
        ok: true,
        value: this.rowToSubAccount(row),
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Get sub-account by ID
   */
  async getSubAccount(id: string): Promise<Result<SubAccount | null>> {
    try {
      const result = await this.pool.query(`
        SELECT * FROM ent_sub_accounts WHERE id = $1
      `, [id]);

      if (result.rows.length === 0) {
        return { ok: true, value: null };
      }

      return { ok: true, value: this.rowToSubAccount(result.rows[0]) };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * List sub-accounts for parent
   */
  async listSubAccounts(
    parentAccountId: string,
    includeDeleted: boolean = false
  ): Promise<Result<SubAccount[]>> {
    try {
      let query = `
        SELECT * FROM ent_sub_accounts
        WHERE parent_account_id = $1
      `;

      if (!includeDeleted) {
        query += ` AND status != 'deleted'`;
      }

      query += ' ORDER BY created_at DESC';

      const result = await this.pool.query(query, [parentAccountId]);

      return {
        ok: true,
        value: result.rows.map(row => this.rowToSubAccount(row)),
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Update sub-account
   */
  async updateSubAccount(
    id: string,
    updates: Partial<{
      name: string;
      email: string;
      status: SubAccountStatus;
      volumeLimit: number;
      settings: Partial<SubAccountSettings>;
    }>
  ): Promise<Result<SubAccount>> {
    try {
      // Get current sub-account
      const current = await this.getSubAccount(id);
      if (!current.ok) return { ok: false, error: current.error };
      if (!current.value) {
        return { ok: false, error: new Error('Sub-account not found') };
      }

      const mergedSettings = updates.settings
        ? { ...current.value.settings, ...updates.settings }
        : current.value.settings;

      const result = await this.pool.query(`
        UPDATE ent_sub_accounts SET
          name = COALESCE($2, name),
          email = COALESCE($3, email),
          status = COALESCE($4, status),
          volume_limit = COALESCE($5, volume_limit),
          settings = $6,
          updated_at = NOW()
        WHERE id = $1
        RETURNING *
      `, [
        id,
        updates.name,
        updates.email,
        updates.status,
        updates.volumeLimit,
        JSON.stringify(mergedSettings),
      ]);

      return { ok: true, value: this.rowToSubAccount(result.rows[0]) };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Suspend sub-account
   */
  async suspendSubAccount(id: string, reason?: string): Promise<Result<void>> {
    try {
      await this.pool.query(`
        UPDATE ent_sub_accounts SET
          status = $2,
          suspended_reason = $3,
          suspended_at = NOW(),
          updated_at = NOW()
        WHERE id = $1
      `, [id, SubAccountStatus.SUSPENDED, reason]);

      // Invalidate API keys
      await this.redis.del(`subaccount:keys:${id}`);

      return { ok: true, value: undefined };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Delete sub-account (soft delete)
   */
  async deleteSubAccount(id: string): Promise<Result<void>> {
    try {
      await this.pool.query(`
        UPDATE ent_sub_accounts SET
          status = $2,
          deleted_at = NOW(),
          updated_at = NOW()
        WHERE id = $1
      `, [id, SubAccountStatus.DELETED]);

      return { ok: true, value: undefined };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Get aggregate statistics for all sub-accounts
   */
  async getAggregateStats(
    parentAccountId: string,
    startDate: Date,
    endDate: Date
  ): Promise<Result<SubAccountStats[]>> {
    try {
      const result = await this.pool.query(`
        SELECT 
          s.id,
          s.name,
          COALESCE(stats.emails_sent, 0) as emails_sent,
          COALESCE(stats.emails_delivered, 0) as emails_delivered,
          COALESCE(stats.emails_bounced, 0) as emails_bounced,
          COALESCE(stats.emails_opened, 0) as emails_opened,
          COALESCE(stats.emails_clicked, 0) as emails_clicked,
          s.volume_used,
          s.volume_limit
        FROM ent_sub_accounts s
        LEFT JOIN (
          SELECT 
            sub_account_id,
            SUM(CASE WHEN event_type = 'sent' THEN 1 ELSE 0 END) as emails_sent,
            SUM(CASE WHEN event_type = 'delivered' THEN 1 ELSE 0 END) as emails_delivered,
            SUM(CASE WHEN event_type = 'bounced' THEN 1 ELSE 0 END) as emails_bounced,
            SUM(CASE WHEN event_type = 'opened' THEN 1 ELSE 0 END) as emails_opened,
            SUM(CASE WHEN event_type = 'clicked' THEN 1 ELSE 0 END) as emails_clicked
          FROM ent_sub_account_events
          WHERE created_at BETWEEN $2 AND $3
          GROUP BY sub_account_id
        ) stats ON s.id = stats.sub_account_id
        WHERE s.parent_account_id = $1 AND s.status != 'deleted'
        ORDER BY stats.emails_sent DESC NULLS LAST
      `, [parentAccountId, startDate, endDate]);

      return {
        ok: true,
        value: result.rows.map(row => ({
          id: row.id,
          name: row.name,
          emailsSent: parseInt(row.emails_sent, 10),
          emailsDelivered: parseInt(row.emails_delivered, 10),
          emailsBounced: parseInt(row.emails_bounced, 10),
          emailsOpened: parseInt(row.emails_opened, 10),
          emailsClicked: parseInt(row.emails_clicked, 10),
          volumeUsed: row.volume_used,
          volumeLimit: row.volume_limit,
          volumePercentage: row.volume_limit > 0 
            ? Math.round((row.volume_used / row.volume_limit) * 100)
            : 0,
        })),
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Allocate volume to sub-account
   */
  async allocateVolume(
    subAccountId: string,
    volume: number,
    periodStart: Date,
    periodEnd: Date
  ): Promise<Result<VolumeAllocation>> {
    try {
      const result = await this.pool.query(`
        INSERT INTO ent_volume_allocations (
          id, sub_account_id, allocated_volume, used_volume,
          period_start, period_end, created_at
        ) VALUES ($1, $2, $3, 0, $4, $5, NOW())
        ON CONFLICT (sub_account_id, period_start) DO UPDATE SET
          allocated_volume = $3
        RETURNING *
      `, [uuidv4(), subAccountId, volume, periodStart, periodEnd]);

      const row = result.rows[0];
      return {
        ok: true,
        value: {
          subAccountId: row.sub_account_id,
          allocatedVolume: row.allocated_volume,
          usedVolume: row.used_volume,
          remainingVolume: row.allocated_volume - row.used_volume,
          periodStart: row.period_start,
          periodEnd: row.period_end,
        },
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Track volume usage
   */
  async trackVolumeUsage(subAccountId: string, count: number = 1): Promise<Result<void>> {
    try {
      // Update sub-account volume
      await this.pool.query(`
        UPDATE ent_sub_accounts SET
          volume_used = volume_used + $2,
          updated_at = NOW()
        WHERE id = $1
      `, [subAccountId, count]);

      // Update current period allocation
      await this.pool.query(`
        UPDATE ent_volume_allocations SET
          used_volume = used_volume + $2
        WHERE sub_account_id = $1
        AND period_start <= NOW()
        AND period_end >= NOW()
      `, [subAccountId, count]);

      // Update Redis counter
      const dateKey = new Date().toISOString().split('T')[0];
      await this.redis.hincrby(`subaccount:volume:${subAccountId}`, dateKey, count);

      return { ok: true, value: undefined };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Check if sub-account has remaining volume
   */
  async checkVolumeLimit(subAccountId: string): Promise<Result<{ allowed: boolean; remaining: number }>> {
    try {
      const result = await this.pool.query(`
        SELECT volume_limit, volume_used FROM ent_sub_accounts
        WHERE id = $1
      `, [subAccountId]);

      if (result.rows.length === 0) {
        return { ok: false, error: new Error('Sub-account not found') };
      }

      const { volume_limit, volume_used } = result.rows[0];

      // If no limit set (0), allow unlimited
      if (volume_limit === 0) {
        return { ok: true, value: { allowed: true, remaining: -1 } };
      }

      const remaining = volume_limit - volume_used;
      return {
        ok: true,
        value: {
          allowed: remaining > 0,
          remaining: Math.max(0, remaining),
        },
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Create API key for sub-account
   */
  async createAPIKey(
    subAccountId: string,
    name: string,
    scopes: string[],
    expiresAt?: Date
  ): Promise<Result<{ key: string; apiKey: SubAccountAPIKey }>> {
    try {
      const id = uuidv4();
      const rawKey = `sak_${uuidv4().replace(/-/g, '')}`;
      const keyPrefix = rawKey.substring(0, 12);
      const keyHash = require('crypto').createHash('sha256').update(rawKey).digest('hex');

      await this.pool.query(`
        INSERT INTO ent_sub_account_api_keys (
          id, sub_account_id, name, key_prefix, key_hash,
          scopes, expires_at, created_at
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, NOW())
      `, [id, subAccountId, name, keyPrefix, keyHash, scopes, expiresAt]);

      const apiKey: SubAccountAPIKey = {
        id,
        subAccountId,
        name,
        keyPrefix,
        keyHash,
        scopes,
        expiresAt,
        createdAt: new Date(),
      };

      // Invalidate cache
      await this.redis.del(`subaccount:keys:${subAccountId}`);

      return {
        ok: true,
        value: { key: rawKey, apiKey },
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Validate sub-account API key
   */
  async validateAPIKey(key: string): Promise<Result<{ subAccountId: string; scopes: string[] } | null>> {
    try {
      const keyHash = require('crypto').createHash('sha256').update(key).digest('hex');

      const result = await this.pool.query(`
        SELECT k.*, s.status as account_status
        FROM ent_sub_account_api_keys k
        JOIN ent_sub_accounts s ON k.sub_account_id = s.id
        WHERE k.key_hash = $1
        AND (k.expires_at IS NULL OR k.expires_at > NOW())
        AND s.status = 'active'
      `, [keyHash]);

      if (result.rows.length === 0) {
        return { ok: true, value: null };
      }

      const row = result.rows[0];

      // Update last used
      await this.pool.query(`
        UPDATE ent_sub_account_api_keys SET last_used_at = NOW() WHERE id = $1
      `, [row.id]);

      return {
        ok: true,
        value: {
          subAccountId: row.sub_account_id,
          scopes: row.scopes,
        },
      };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  /**
   * Record sub-account event
   */
  async recordEvent(
    subAccountId: string,
    eventType: string,
    messageId?: string,
    metadata?: Record<string, any>
  ): Promise<Result<void>> {
    try {
      await this.pool.query(`
        INSERT INTO ent_sub_account_events (
          id, sub_account_id, event_type, message_id, metadata, created_at
        ) VALUES ($1, $2, $3, $4, $5, NOW())
      `, [uuidv4(), subAccountId, eventType, messageId, JSON.stringify(metadata || {})]);

      return { ok: true, value: undefined };
    } catch (error) {
      return {
        ok: false,
        error: error instanceof Error ? error : new Error(String(error)),
      };
    }
  }

  private rowToSubAccount(row: any): SubAccount {
    return {
      id: row.id,
      parentAccountId: row.parent_account_id,
      name: row.name,
      slug: row.slug,
      email: row.email,
      status: row.status as SubAccountStatus,
      volumeLimit: row.volume_limit,
      volumeUsed: row.volume_used,
      settings: row.settings,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    };
  }
}

File: services/mail-server/crates/enterprise/src/sub_accounts.rs
use sha2::{Sha256, Digest};
use sqlx::PgPool;
use tracing::info;
use uuid::Uuid;

use crate::config::VolumeAllocationMode;
use crate::types::*;

/// Sub-Accounts Service: agency/reseller sub-accounts, volume allocation, API keys
pub struct SubAccountService {
    db: PgPool,
    max_sub_accounts: i32,
    volume_allocation_mode: VolumeAllocationMode,
}

impl SubAccountService {
    pub fn new(db: PgPool, max_sub_accounts: i32, volume_allocation_mode: VolumeAllocationMode) -> Self {
        Self { db, max_sub_accounts, volume_allocation_mode }
    }

    /// Create a sub-account
    pub async fn create(
        &self, parent_id: Uuid, name: &str, email: Option<&str>,
        domain: Option<&str>, plan: Option<&str>, volume_limit: Option<i64>,
        inherit_parent_settings: bool,
    ) -> Result<ApiResult<SubAccount>, String> {
        // Check sub-account limit
        let count: (i64,) = sqlx::query_as(
            "SELECT COUNT(*) FROM ent_sub_accounts WHERE parent_id = $1"
        )
        .bind(parent_id)
        .fetch_one(&self.db)
        .await
        .map_err(|e| format!("Count sub-accounts: {e}"))?;

        if count.0 >= self.max_sub_accounts as i64 {
            return Ok(ApiResult::err(
                format!("Maximum sub-accounts ({}) reached", self.max_sub_accounts),
                "QUOTA_EXCEEDED",
            ));
        }

        let id = Uuid::new_v4();
        let row = sqlx::query_as::<_, SubAccount>(
            "INSERT INTO ent_sub_accounts (id, parent_id, name, status, email, domain, plan, volume_limit, volume_used, inherit_parent_settings, created_at, updated_at)
             VALUES ($1,$2,$3,'active',$4,$5,$6,$7,0,$8,NOW(),NOW())
             RETURNING *"
        )
        .bind(id).bind(parent_id).bind(name).bind(email)
        .bind(domain).bind(plan).bind(volume_limit).bind(inherit_parent_settings)
        .fetch_one(&self.db)
        .await
        .map_err(|e| format!("Create sub-account: {e}"))?;

        info!(parent_id = %parent_id, sub_id = %id, name = name, "Sub-account created");
        Ok(ApiResult::ok(row))
    }

    /// Get a sub-account by ID
    pub async fn get(&self, id: Uuid) -> Result<ApiResult<SubAccount>, String> {
        let row = sqlx::query_as::<_, SubAccount>(
            "SELECT * FROM ent_sub_accounts WHERE id = $1"
        )
        .bind(id)
        .fetch_optional(&self.db)
        .await
        .map_err(|e| format!("Get sub-account: {e}"))?;

        match row {
            Some(r) => Ok(ApiResult::ok(r)),
            None => Ok(ApiResult::err("Sub-account not found", "NOT_FOUND")),
        }
    }

    /// List sub-accounts for a parent
    pub async fn list(
        &self, parent_id: Uuid, status: Option<&str>, limit: i64, offset: i64,
    ) -> Result<ApiResult<Vec<SubAccount>>, String> {
        let rows = if let Some(s) = status {
            sqlx::query_as::<_, SubAccount>(
                "SELECT * FROM ent_sub_accounts WHERE parent_id = $1 AND status = $2 ORDER BY created_at DESC LIMIT $3 OFFSET $4"
            )
            .bind(parent_id).bind(s).bind(limit).bind(offset)
            .fetch_all(&self.db)
            .await
        } else {
            sqlx::query_as::<_, SubAccount>(
                "SELECT * FROM ent_sub_accounts WHERE parent_id = $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3"
            )
            .bind(parent_id).bind(limit).bind(offset)
            .fetch_all(&self.db)
            .await
        }.map_err(|e| format!("List sub-accounts: {e}"))?;

        Ok(ApiResult::ok(rows))
    }

    /// Update a sub-account
    pub async fn update(
        &self, id: Uuid, name: Option<&str>, email: Option<&str>,
        volume_limit: Option<i64>, settings: Option<serde_json::Value>,
    ) -> Result<ApiResult<SubAccount>, String> {
        let row = sqlx::query_as::<_, SubAccount>(
            "UPDATE ent_sub_accounts SET
             name = COALESCE($2, name),
             email = COALESCE($3, email),
             volume_limit = COALESCE($4, volume_limit),
             settings = COALESCE($5, settings),
             updated_at = NOW()
             WHERE id = $1 RETURNING *"
        )
        .bind(id).bind(name).bind(email).bind(volume_limit).bind(&settings)
        .fetch_optional(&self.db)
        .await
        .map_err(|e| format!("Update sub-account: {e}"))?;

        match row {
            Some(r) => Ok(ApiResult::ok(r)),
            None => Ok(ApiResult::err("Sub-account not found", "NOT_FOUND")),
        }
    }

    /// Suspend a sub-account
    pub async fn suspend(&self, id: Uuid, reason: Option<&str>) -> Result<ApiResult<SubAccount>, String> {
        let row = sqlx::query_as::<_, SubAccount>(
            "UPDATE ent_sub_accounts SET status = 'suspended', updated_at = NOW() WHERE id = $1 RETURNING *"
        )
        .bind(id)
        .fetch_optional(&self.db)
        .await
        .map_err(|e| format!("Suspend sub-account: {e}"))?;

        match row {
            Some(r) => {
                info!(id = %id, reason = ?reason, "Sub-account suspended");
                Ok(ApiResult::ok(r))
            }
            None => Ok(ApiResult::err("Sub-account not found", "NOT_FOUND")),
        }
    }

    /// Delete a sub-account
    pub async fn delete(&self, id: Uuid) -> Result<ApiResult<serde_json::Value>, String> {
        let result = sqlx::query("DELETE FROM ent_sub_accounts WHERE id = $1")
            .bind(id)
            .execute(&self.db)
            .await
            .map_err(|e| format!("Delete sub-account: {e}"))?;

        if result.rows_affected() == 0 {
            Ok(ApiResult::err("Sub-account not found", "NOT_FOUND"))
        } else {
            Ok(ApiResult::ok(serde_json::json!({"deleted": true})))
        }
    }

    /// Get aggregate stats for all sub-accounts of a parent
    pub async fn get_stats(&self, parent_id: Uuid) -> Result<ApiResult<SubAccountStats>, String> {
        let row: (i64, i64, Option<i64>, Option<i64>) = sqlx::query_as(
            "SELECT COUNT(*), COUNT(*) FILTER (WHERE status = 'active'),
             SUM(volume_used)::bigint, SUM(volume_limit)::bigint
             FROM ent_sub_accounts WHERE parent_id = $1"
        )
        .bind(parent_id)
        .fetch_one(&self.db)
        .await
        .map_err(|e| format!("Get sub-account stats: {e}"))?;

        Ok(ApiResult::ok(SubAccountStats {
            total: row.0,
            active: row.1,
            total_volume_used: row.2.unwrap_or(0),
            total_volume_limit: row.3.unwrap_or(0),
        }))
    }

    /// Create an API key for a sub-account
    pub async fn create_api_key(
        &self, sub_account_id: Uuid, name: &str, permissions: Option<Vec<String>>,
        rate_limit: Option<i32>,
    ) -> Result<ApiResult<serde_json::Value>, String> {
        let raw_key = crate::sso::generate_random_token(32);
        let key_prefix = &raw_key[..8];
        let key_hash = sha256_hex(&raw_key);

        let id = Uuid::new_v4();
        let _row = sqlx::query_as::<_, SubAccountApiKey>(
            "INSERT INTO ent_sub_account_api_keys (id, sub_account_id, key_hash, key_prefix, name, permissions, rate_limit, revoked, created_at)
             VALUES ($1,$2,$3,$4,$5,$6,$7,false,NOW())
             RETURNING *"
        )
        .bind(id).bind(sub_account_id).bind(&key_hash).bind(key_prefix)
        .bind(name).bind(&permissions).bind(rate_limit)
        .fetch_one(&self.db)
        .await
        .map_err(|e| format!("Create API key: {e}"))?;

        info!(sub_account_id = %sub_account_id, "API key created");
        // Return the raw key only once — it cannot be retrieved later
        Ok(ApiResult::ok(serde_json::json!({
            "id": id,
            "key": raw_key,
            "key_prefix": key_prefix,
            "name": name,
        })))
    }

    /// Check volume allocation for a sub-account
    pub fn check_volume_allowed(&self, volume_used: i64, volume_limit: Option<i64>, parent_headroom: i64) -> bool {
        check_volume_logic(&self.volume_allocation_mode, volume_used, volume_limit, parent_headroom)
    }
}

/// Pure function: volume check logic (extracted for testability)
pub fn check_volume_logic(
    mode: &VolumeAllocationMode, volume_used: i64, volume_limit: Option<i64>, _parent_headroom: i64,
) -> bool {
    match mode {
        VolumeAllocationMode::Fixed => {
            volume_limit.map(|l| volume_used < l).unwrap_or(true)
        }
        VolumeAllocationMode::Shared => {
            // Shared mode: no hard per-account limit
            true
        }
        VolumeAllocationMode::Burst => {
            // Burst mode: allow up to 120% of limit if parent has headroom
            volume_limit.map(|l| {
                let burst_limit = (l as f64 * 1.2) as i64;
                volume_used < burst_limit
            }).unwrap_or(true)
        }
    }
}

/// Compute SHA-256 hex digest
pub fn sha256_hex(input: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(input.as_bytes());
    hex::encode(hasher.finalize())
}

// ── Tests ──────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;

    #[test]
    fn test_sha256_hex() {
        let hash = sha256_hex("test");
        assert_eq!(hash.len(), 64);
        assert_eq!(hash, "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08");
    }

    #[test]
    fn test_sha256_deterministic() {
        assert_eq!(sha256_hex("hello"), sha256_hex("hello"));
    }

    #[test]
    fn test_volume_fixed_under_limit() {
        assert!(check_volume_logic(&VolumeAllocationMode::Fixed, 500, Some(1000), 0));
    }

    #[test]
    fn test_volume_fixed_over_limit() {
        assert!(!check_volume_logic(&VolumeAllocationMode::Fixed, 1000, Some(1000), 0));
    }

    #[test]
    fn test_volume_shared_always_allowed() {
        assert!(check_volume_logic(&VolumeAllocationMode::Shared, 999999, Some(100), 0));
    }

    #[test]
    fn test_volume_burst_within_120_percent() {
        assert!(check_volume_logic(&VolumeAllocationMode::Burst, 1100, Some(1000), 5000));
    }

    #[test]
    fn test_volume_burst_over_120_percent() {
        assert!(!check_volume_logic(&VolumeAllocationMode::Burst, 1200, Some(1000), 5000));
    }

    #[test]
    fn test_volume_no_limit() {
        assert!(check_volume_logic(&VolumeAllocationMode::Fixed, 999999, None, 0));
    }

    #[test]
    fn test_sub_account_stats_serialization() {
        let stats = SubAccountStats {
            total: 10,
            active: 8,
            total_volume_used: 50000,
            total_volume_limit: 100000,
        };
        let json = serde_json::to_value(&stats).unwrap();
        assert_eq!(json["total"], 10);
        assert_eq!(json["active"], 8);
    }

    #[test]
    fn test_sub_account_serialization() {
        let sa = SubAccount {
            id: Uuid::new_v4(),
            parent_id: Uuid::new_v4(),
            name: "Agency Client 1".into(),
            status: "active".into(),
            email: Some("client@agency.com".into()),
            domain: Some("client.agency.com".into()),
            plan: Some("business".into()),
            volume_limit: Some(50000),
            volume_used: 12000,
            inherit_parent_settings: true,
            settings: None,
            metadata: None,
            created_at: Some(Utc::now()),
            updated_at: Some(Utc::now()),
        };
        let json = serde_json::to_value(&sa).unwrap();
        assert_eq!(json["name"], "Agency Client 1");
        assert_eq!(json["volume_used"], 12000);
    }
}
